### PR TITLE
fix bug in Serialize (#74447)

### PR DESCRIPTION
**Pull request by SmirnovGrigory**
### Details:
- *When using ngraph::pass::Serialize there are some side effects to the underlying model and some tests from gnaFuncTests are failing*
-  *it was found that Serialize changes pads in nGraph function*
- *Because of it we need to copy function to let Serialize work with copy*
- *Added simple test to check that pads doesn't change now*

### Tickets:
 - *74447*

### CODE REVIEW

**Comment by ilya-lavrenov** [comment]
> it was found that Serialize changes pads in nGraph function

Why does it do it? can it be fixed w/o explicit copy of the graph?
Does it change pads of original model or serialized model has changed pads as well?

File: src/core/src/pass/serialize.cpp [hunk 1]
```diff
@@ -1012,7 +1012,8 @@ void serializeFunc(std::ostream& xml_file,
 }  // namespace
 
 namespace ov {
-bool pass::Serialize::run_on_model(const std::shared_ptr<ngraph::Function>& f) {
+bool pass::Serialize::run_on_model(const std::shared_ptr<ngraph::Function>& f_orig) {
+    auto f = ngraph::clone_function(*f_orig);
```
Comment (GlebKazantaev) [inline src/core/src/pass/serialize.cpp:1016]: Add a comment with explanations why this copy is needed here.

File: src/core/tests/pass/serialization/serialize.cpp [hunk 2]
```diff
@@ -51,6 +53,32 @@ TEST_P(SerializationTest, CompareFunctions) {
...
+                                                          ngraph::op::PadType::AUTO);
+    auto f = std::make_shared<ov::Model>(ov::NodeVector{conv}, ov::ParameterVector{input});
+
+    ov::pass::Manager manager;
+    manager.register_pass<ngraph::pass::Serialize>("pads_test.xml", "pads_test.bin");
+    ASSERT_NO_THROW(manager.run_passes(f));
+
+    auto pad = ov::CoordinateDiff{1, 1};
+    ASSERT_EQ(conv->get_pads_begin(), pad);
+    ASSERT_EQ(conv->get_pads_end(), pad);
```
Comment (GlebKazantaev) [inline src/core/tests/pass/serialization/serialize.cpp:76]: It is better to use `CompareFunctions` here. So you first copy function, then call Serialize and then compare functions.
Also I would suggest to extend other existing tests with this check.

**Comment by SmirnovGrigory** [comment]
> > it was found that Serialize changes pads in nGraph function
> 
> Why does it do it? can it be fixed w/o explicit copy of the graph? Does it change pads of original model or serialized model has changed pads as well?

Serialization changes paddings of convolutions with autopad. Also we we write to IR pads {0,0} if convolution has autopad. My colleagues says that copying ngraph function would be the best solution

File: src/core/src/pass/serialize.cpp (hunk 1 above)
Comment (SmirnovGrigory) [reply src/core/src/pass/serialize.cpp:1019]: done

File: src/core/tests/CMakeLists.txt [hunk 3]
```diff
@@ -75,10 +75,6 @@ set(SRC
     partial_shape.cpp
     pass_config.cpp
     pass_manager.cpp
-    pass/serialization/cleanup.cpp
```
Comment (ilyachur) [inline src/core/tests/CMakeLists.txt:78]: Serialize transformation is a part of public API and it is a part of core component. Why did you move these tests to inference_engine func tests? Please revert these changes.

File: src/tests/functional/inference_engine/transformations/deterministicity.cpp [hunk 4]
```diff
@@ -9,8 +9,8 @@
 #include "openvino/pass/serialize.hpp"
 #include "openvino/util/file_util.hpp"
 #include "read_ir.hpp"
-#include "util/graph_comparator.hpp"
-#include "util/test_common.hpp"
+#include <graph_comparator.hpp>
```
Comment (ilyachur) [inline src/tests/functional/inference_engine/transformations/deterministicity.cpp:12]: Please revert these changes

**Review by ilyachur — CHANGES_REQUESTED** [review]
Please revert all changes in tests.

File: src/tests/functional/inference_engine/transformations/deterministicity.cpp (hunk 4 above)
Comment (SmirnovGrigory) [reply src/tests/functional/inference_engine/transformations/deterministicity.cpp]: done

File: src/core/tests/CMakeLists.txt (hunk 3 above)
Comment (SmirnovGrigory) [reply src/core/tests/CMakeLists.txt]: done

File: src/core/tests/pass/serialization/serialize.cpp [hunk 5]
```diff
@@ -7,6 +7,8 @@
 #include <gtest/gtest.h>
 
 #include <fstream>
+#include <ngraph/pass/manager.hpp>
+#include <ngraph/pass/serialize.hpp>
```
Comment (ilyachur) [inline src/core/tests/pass/serialization/serialize.cpp:11]:
```suggestion
#include "ngraph/pass/manager.hpp"
#include "ngraph/pass/serialize.hpp"
```

File: src/core/tests/pass/serialization/serialize.cpp [hunk 6]
```diff
@@ -40,15 +42,17 @@ class SerializationTest : public ov::test::TestsCommon, public testing::WithPara
 };
 
 TEST_P(SerializationTest, CompareFunctions) {
+    auto orig = ov::test::readModel(m_model_path, m_binary_path);
     auto expected = ov::test::readModel(m_model_path, m_binary_path);
```
Comment (ilyachur) [inline src/core/tests/pass/serialization/serialize.cpp:46]: Looks like you read the same model two times.
You can just clone_model

File: src/core/src/pass/serialize.cpp (hunk 1 above)
Comment (ilyachur) [inline src/core/src/pass/serialize.cpp:1016]:
```suggestion
    auto f = ov::clone_model(*f_orig);
```

File: src/core/tests/pass/serialization/serialize.cpp (hunk 5 above)
Comment (SmirnovGrigory) [reply src/core/tests/pass/serialization/serialize.cpp]: done

File: src/core/src/pass/serialize.cpp (hunk 1 above)
Comment (SmirnovGrigory) [reply src/core/src/pass/serialize.cpp]: done

File: src/core/tests/pass/serialization/serialize.cpp (hunk 6 above)
Comment (SmirnovGrigory) [reply src/core/tests/pass/serialization/serialize.cpp]: done

**Review by ilyachur — APPROVED** [review]

**Review by nkogteva — APPROVED** [review]

File: src/core/include/openvino/core/node_input.hpp [hunk 7]
```diff
@@ -31,7 +31,12 @@ class OPENVINO_API Input<Node> {
     /// \param node Pointer to the node for the input handle.
     /// \param index The index of the input.
     Input(Node* node, size_t index);
-
+    /// \brief Constructs a Input.
+    /// \param node A `shared_ptr` to the node for the input handle.
+    /// \param index The index of the input.
+    Input(const std::shared_ptr<Node>& node, size_t index);
```
Comment (GlebKazantaev) [inline src/core/include/openvino/core/node_input.hpp:37]: Why we need this if we have Input(Node* node, size_t index) ?

File: src/core/src/graph_util.cpp [hunk 8]
```diff
@@ -245,6 +245,12 @@ std::vector<std::shared_ptr<ngraph::Node>> ngraph::clone_nodes(const std::vector
                 new_output.get_rt_info() = output_rt_info;
             }
 
+            for (auto input : node->inputs()) {
+                const auto& output_rt_info = input.get_rt_info();
+                auto new_input = input.for_node(cloned_node);
```
Comment (GlebKazantaev) [inline src/core/src/graph_util.cpp:250]: From my point of view `for_node` is not required as we can use `cloned_node->input(input.get_index())`

File: src/core/src/node_input.cpp [hunk 9]
```diff
@@ -9,6 +9,12 @@
 namespace ov {
 Input<Node>::Input(Node* node, size_t index) : m_node(node), m_index(index) {}
 
+Input<Node>::Input(const std::shared_ptr<Node>& node, size_t index) : m_node(node.get()), m_index(index) {}
```
Comment (GlebKazantaev) [inline src/core/src/node_input.cpp:12]: Input<Node> class doesn't extend life for shared_ptr<Node>,  so  only Output<Node> has c-tor which takes shared_ptr as input.

File: src/core/src/graph_util.cpp (hunk 8 above)
Comment (SmirnovGrigory) [reply src/core/src/graph_util.cpp]: Ok! changed for_node to cloned_node->input(input.get_index())

**Review by GlebKazantaev — APPROVED** [review]
